From bd07f7c7b3f80710762578f85c325c0be5d8c779 Mon Sep 17 00:00:00 2001
From: Noam Lessner <nlessner@habana.ai>
Date: Tue, 18 Feb 2020 13:48:14 +0200
Subject: [PATCH] [SW-8229] goya2: save efuse in scratchpad_sram

To keep verification routine independent from global, so arcboot_ram
can also call the same routine without depending on variables or
globals defined exclusively in arcboot_rom.

Change-Id: I8a4f712639828540a0ec4c99c75f8e4f9918b3fe
Signed-off-by: Noam Lessner <nlessner@habana.ai>
---

diff --git a/app/arcboot_rom.c b/app/arcboot_rom.c
index 217575c..fdb132e 100644
--- a/app/arcboot_rom.c
+++ b/app/arcboot_rom.c
@@ -152,6 +152,10 @@
 		      &arcboot_rom_global.config_fuse) == -1)
 		arcboot_rom_global.config_fuse = EFUSE_SYSCFG_DEFAULT;
 
+	/* store key hashes in Scratch PAD Sram */
+	sp_sram_writel(arcboot_rom_global.config_fuse,
+		       SECURITY_EFUSE_SYS_CFG_OFFSET);
+
 	/* TODO: check in pldm if we have access, if yes get waves */
 	if (arcboot_rom_global.config_fuse & EFUSE_SYSCFG_DISABLE_DEBUG_IF_MASK)
 		security_disable_debug_if();
diff --git a/common/secureboot.c b/common/secureboot.c
index f84a851..622baf4 100644
--- a/common/secureboot.c
+++ b/common/secureboot.c
@@ -37,6 +37,7 @@
 	int key_scan;
 	unsigned char block_buf[SECUREBOOT_HASH_BLOCK_CALC_BUF];
 	unsigned char test_phase0a_hash[SECURITY_KEY_HASH_ARRAY_BIT_SIZE];
+	uint32_t config_fuse;
 	uint16_t test_phase0b_data_len = image_size;
 	uint32_t test_phase0b_sig_len = SIG_SIZE;
 	uint32_t test_public_key_len = KEY_SIZE;
@@ -48,10 +49,9 @@
 					 test_public_key_len;
 	uint32_t fw_version = base_readl((uint32_t)test_phase0b_data);
 
-	/* TODO: remove any global or ROM specific rutines */
 	/* check efuse for security validation enable */
-	if ((arcboot_rom_global.config_fuse &
-	     EFUSE_SYSCFG_ENABLE_VALIDATION_MASK) == 0)
+	config_fuse = sp_sram_readl(SECURITY_EFUSE_SYS_CFG_OFFSET);
+	if ((config_fuse & EFUSE_SYSCFG_ENABLE_VALIDATION_MASK) == 0)
 		/* if efuse isn't set, check status of security bypass
 		 * boot strap if enabled bypass security validation.
 		 */
diff --git a/common/security.h b/common/security.h
index 2c21236..9a3c6ff 100644
--- a/common/security.h
+++ b/common/security.h
@@ -99,6 +99,8 @@
 #define SECURITY_EFUSE_KEYS_OFFSET		16
 /* 16 bytes */
 #define SECURITY_EFUSE_KEYS_REVOKE_OFFSET	176
+/* 4 bytes */
+#define SECURITY_EFUSE_SYS_CFG_OFFSET		192
 
 void security_disable_debug_if(void);
 void security_debug_if_secure(void);
